Remove BinaryReader and BinaryWriter references from ZipArchive

edwardneal opened the pull request:
Relates to #83909, #31460.

This removes all references to BinaryReader and BinaryWriter from ZipArchive and ZipArchiveEntry. It also changes the way that the ZIP central directory header is read from the source stream and makes one tweak to the way that the EOCD header is detected.

I've removed BinaryReader and BinaryWriter for three reasons:
* They were being used to read and write to the stream field-by-field - so every file could generate nearly two dozen individual writes to the backing stream.
* There are no async implementations on these two objects, so these need to be implemented (or their usage removed) in order to add any future async support for ZipArchive.
* Removes some allocations & GC pressure.

The second change is to adjust the way that the ZIP central directory header is read. Previously, this was read from the source stream file-by-file. This PR now reads from the source in 4KB blocks and tries to read the headers from there. This is much faster. I've chosen not to implement it when writing the CD headers because they contain dynamic data and I didn't want to copy buffers around; I'm open to doing so.

The detection of the end-of-central-directory header is very similar too. It was already doing something similar, but with only 16 bytes at a time. I've tweaked this to read 4KB block instead, and changed the way it searches for the EOCD signature to use an intrinsic rather than byte-by-byte bit shuffling.

In both cases, I've picked 4KB because it feels like a small enough buffer to not make a massive difference to wait times, and it aligns with the 4KB buffer on FileStream (which I imagine would be the most common use case.)

There are performance improvements across the board. To benchmark this, I used a test wrapping stream which simulates the worst case - an Xms Thread.Sleep on every Read and Write call. Results are below, but in short:

#### Reads ####
* 12-13% reduction in execution time as a baseline, rising to 18-19% as the number of entries in the archive increases.
* 33-36% reduction in memory usage
* When latency is introduced, as the number of entries in the archive increases, the reduction in execution time becomes more pronounced - 99.6%

#### Creation ####
* Execution time is almost identical assuming no latency. As latency increases, the execution time reduces by around 82%
* 10% reduction in memory usage

<details>
<summary>Benchmark results - Reads</summary>

| Method    | Job      | Runtime  | NumberOfFiles | LatencyMS | Mean              | Error          | StdDev         | Median            | Ratio | RatioSD | Gen0   | Gen1   | Allocated | Alloc Ratio |
|---------- |--------- |--------- |-------------- |---------- |------------------:|---------------:|---------------:|------------------:|------:|--------:|-------:|-------:|----------:|------------:|
| **Benchmark** | **Baseline** | **.NET 8.0** | **2**             | **-1**        |          **1.280 μs** |      **0.0178 μs** |      **0.0338 μs** |          **1.269 μs** |  **1.00** |    **0.00** | **0.4482** |      **-** |   **1.84 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 2             | -1        |          1.131 μs |      0.0104 μs |      0.0081 μs |          1.133 μs |  0.87 |    0.03 | 0.2861 |      - |   1.17 KB |        0.64 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **2**             | **0**         |         **21.412 μs** |      **0.5187 μs** |      **1.5211 μs** |         **21.083 μs** |  **1.00** |    **0.00** | **0.4578** |      **-** |   **1.88 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 2             | 0         |          2.871 μs |      0.0318 μs |      0.0265 μs |          2.871 μs |  0.13 |    0.01 | 0.2937 |      - |   1.21 KB |        0.65 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **2**             | **1**         |    **716,595.380 μs** |  **2,212.8744 μs** |  **2,069.9241 μs** |    **716,359.000 μs** |  **1.00** |    **0.00** |      **-** |      **-** |   **2.64 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 2             | 1         |     62,065.810 μs |    280.6085 μs |    248.7521 μs |     62,120.450 μs |  0.09 |    0.00 |      - |      - |   1.29 KB |        0.49 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **2**             | **5**         |    **716,582.567 μs** |  **2,484.5321 μs** |  **2,324.0329 μs** |    **716,860.600 μs** |  **1.00** |    **0.00** |      **-** |      **-** |   **2.64 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 2             | 5         |     62,190.336 μs |    212.1850 μs |    198.4780 μs |     62,244.822 μs |  0.09 |    0.00 |      - |      - |   1.29 KB |        0.49 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **10**            | **-1**        |          **3.925 μs** |      **0.0296 μs** |      **0.0247 μs** |          **3.922 μs** |  **1.00** |    **0.00** | **1.7929** |      **-** |   **7.33 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 10            | -1        |          3.452 μs |      0.0683 μs |      0.0786 μs |          3.424 μs |  0.88 |    0.03 | 1.1864 |      - |   4.85 KB |        0.66 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **10**            | **0**         |         **98.469 μs** |      **5.5739 μs** |     **16.0820 μs** |         **92.865 μs** |  **1.00** |    **0.00** | **1.7090** |      **-** |   **7.37 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 10            | 0         |          5.902 μs |      0.1166 μs |      0.2011 μs |          5.830 μs |  0.07 |    0.01 | 1.1902 |      - |   4.89 KB |        0.66 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **10**            | **1**         |  **2,966,581.267 μs** |  **5,120.4172 μs** |  **4,789.6414 μs** |  **2,967,184.400 μs** |  **1.00** |    **0.00** |      **-** |      **-** |   **8.09 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 10            | 1         |     62,189.399 μs |    283.9015 μs |    265.5616 μs |     62,213.289 μs |  0.02 |    0.00 |      - |      - |   4.97 KB |        0.61 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **10**            | **5**         |  **2,965,424.293 μs** |  **5,345.5963 μs** |  **5,000.2741 μs** |  **2,965,234.300 μs** |  **1.00** |    **0.00** |      **-** |      **-** |   **8.09 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 10            | 5         |     62,184.896 μs |    320.7364 μs |    300.0170 μs |     62,224.356 μs |  0.02 |    0.00 |      - |      - |   4.97 KB |        0.61 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **25**            | **-1**        |          **9.367 μs** |      **0.1231 μs** |      **0.1028 μs** |          **9.356 μs** |  **1.00** |    **0.00** | **4.2114** |      **-** |  **17.22 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 25            | -1        |          7.661 μs |      0.0472 μs |      0.0394 μs |          7.654 μs |  0.82 |    0.01 | 2.7771 |      - |  11.34 KB |        0.66 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **25**            | **0**         |        **189.773 μs** |      **1.5383 μs** |      **1.2010 μs** |        **190.032 μs** |  **1.00** |    **0.00** | **4.1504** |      **-** |  **17.26 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 25            | 0         |          9.558 μs |      0.1294 μs |      0.1210 μs |          9.529 μs |  0.05 |    0.00 | 2.7771 |      - |  11.38 KB |        0.66 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **25**            | **1**         |  **7,184,576.887 μs** |  **6,870.6351 μs** |  **6,426.7964 μs** |  **7,183,295.200 μs** | **1.000** |    **0.00** |      **-** |      **-** |  **17.98 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 25            | 1         |     62,115.901 μs |    231.8962 μs |    216.9159 μs |     62,113.444 μs | 0.009 |    0.00 |      - |      - |  11.46 KB |        0.64 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **25**            | **5**         |  **7,176,659.979 μs** |  **7,168.4172 μs** |  **6,354.6151 μs** |  **7,177,295.900 μs** | **1.000** |    **0.00** |      **-** |      **-** |  **17.98 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 25            | 5         |     62,154.444 μs |    243.2540 μs |    227.5399 μs |     62,233.989 μs | 0.009 |    0.00 |      - |      - |  11.46 KB |        0.64 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **50**            | **-1**        |         **18.166 μs** |      **0.2379 μs** |      **0.2225 μs** |         **18.171 μs** |  **1.00** |    **0.00** | **8.4229** |      **-** |  **34.48 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 50            | -1        |         14.800 μs |      0.1155 μs |      0.1080 μs |         14.796 μs |  0.81 |    0.01 | 5.6152 |      - |  22.95 KB |        0.67 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **50**            | **0**         |        **368.723 μs** |      **3.1490 μs** |      **2.9455 μs** |        **367.986 μs** |  **1.00** |    **0.00** | **8.3008** |      **-** |  **34.52 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 50            | 0         |         17.192 μs |      0.2263 μs |      0.2117 μs |         17.291 μs |  0.05 |    0.00 | 5.6152 | 0.0305 |  22.98 KB |        0.67 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **50**            | **1**         | **14,205,316.860 μs** | **11,442.8220 μs** | **10,703.6228 μs** | **14,203,486.100 μs** | **1.000** |    **0.00** |      **-** |      **-** |  **35.24 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 50            | 1         |     62,229.530 μs |    282.4781 μs |    264.2302 μs |     62,229.311 μs | 0.004 |    0.00 |      - |      - |  23.06 KB |        0.65 |
|           |          |          |               |           |                   |                |                |                   |       |         |        |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **50**            | **5**         | **14,211,658.147 μs** |  **9,950.6110 μs** |  **9,307.8077 μs** | **14,212,125.500 μs** | **1.000** |    **0.00** |      **-** |      **-** |  **35.24 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 50            | 5         |     62,237.983 μs |    341.3432 μs |    319.2927 μs |     62,250.889 μs | 0.004 |    0.00 |      - |      - |  23.06 KB |        0.65 |

</details>

<details>
<summary>Benchmark results - Reads (scaling up by number of ZipArchiveEntry children)</summary>

| Method    | Job      | Runtime  | NumberOfFiles | LatencyMS | Mean          | Error       | StdDev      | Ratio | RatioSD | Gen0      | Gen1     | Gen2     | Allocated  | Alloc Ratio |
|---------- |--------- |--------- |-------------- |---------- |--------------:|------------:|------------:|------:|--------:|----------:|---------:|---------:|-----------:|------------:|
| **Benchmark** | **Baseline** | **.NET 8.0** | **2**             | **-1**        |      **1.276 μs** |   **0.0162 μs** |   **0.0152 μs** |  **1.00** |    **0.00** |    **0.4482** |        **-** |        **-** |    **1.84 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 2             | -1        |      1.120 μs |   0.0116 μs |   0.0129 μs |  0.88 |    0.01 |    0.2861 |        - |        - |    1.17 KB |        0.64 |
|           |          |          |               |           |               |             |             |       |         |           |          |          |            |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **10**            | **-1**        |      **4.527 μs** |   **0.0259 μs** |   **0.0229 μs** |  **1.00** |    **0.00** |    **1.7929** |        **-** |        **-** |    **7.33 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 10            | -1        |      3.964 μs |   0.0315 μs |   0.0295 μs |  0.88 |    0.01 |    1.1826 |        - |        - |    4.85 KB |        0.66 |
|           |          |          |               |           |               |             |             |       |         |           |          |          |            |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **25**            | **-1**        |     **10.897 μs** |   **0.0787 μs** |   **0.0698 μs** |  **1.00** |    **0.00** |    **4.2114** |        **-** |        **-** |   **17.22 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 25            | -1        |      8.824 μs |   0.0499 μs |   0.0466 μs |  0.81 |    0.01 |    2.7771 |        - |        - |   11.34 KB |        0.66 |
|           |          |          |               |           |               |             |             |       |         |           |          |          |            |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **50**            | **-1**        |     **21.540 μs** |   **0.1920 μs** |   **0.1604 μs** |  **1.00** |    **0.00** |    **8.4229** |        **-** |        **-** |   **34.48 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 50            | -1        |     17.341 μs |   0.1089 μs |   0.1019 μs |  0.81 |    0.01 |    5.6152 |        - |        - |   22.95 KB |        0.67 |
|           |          |          |               |           |               |             |             |       |         |           |          |          |            |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **100**           | **-1**        |     **43.626 μs** |   **0.4626 μs** |   **0.6016 μs** |  **1.00** |    **0.00** |   **16.9678** |   **0.1221** |        **-** |   **69.46 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 100           | -1        |     33.674 μs |   0.1899 μs |   0.1683 μs |  0.77 |    0.01 |   11.3525 |        - |        - |   46.59 KB |        0.67 |
|           |          |          |               |           |               |             |             |       |         |           |          |          |            |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **250**           | **-1**        |    **109.428 μs** |   **1.3621 μs** |   **1.2074 μs** |  **1.00** |    **0.00** |   **41.2598** |   **0.1221** |        **-** |  **168.87 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 250           | -1        |     86.518 μs |   1.4735 μs |   1.2305 μs |  0.79 |    0.01 |   27.3438 |   6.7139 |        - |  112.02 KB |        0.66 |
|           |          |          |               |           |               |             |             |       |         |           |          |          |            |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **500**           | **-1**        |    **222.703 μs** |   **1.4561 μs** |   **1.3621 μs** |  **1.00** |    **0.00** |   **83.2520** |   **0.2441** |        **-** |  **340.65 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 500           | -1        |    171.378 μs |   1.2795 μs |   1.0684 μs |  0.77 |    0.01 |   55.4199 |   0.2441 |        - |  227.16 KB |        0.67 |
|           |          |          |               |           |               |             |             |       |         |           |          |          |            |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **1000**          | **-1**        |    **476.601 μs** |   **8.2338 μs** |   **6.8756 μs** |  **1.00** |    **0.00** |  **133.7891** |  **72.2656** |        **-** |  **686.68 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 1000          | -1        |    361.508 μs |   1.1992 μs |   1.0630 μs |  0.76 |    0.01 |   90.8203 |  47.3633 |        - |  459.91 KB |        0.67 |
|           |          |          |               |           |               |             |             |       |         |           |          |          |            |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **10000**         | **-1**        | **12,294.500 μs** | **244.9925 μs** | **547.9609 μs** |  **1.00** |    **0.00** | **1171.8750** | **687.5000** | **265.6250** | **6879.98 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 10000         | -1        | 10,722.270 μs | 212.4330 μs | 504.8696 μs |  0.87 |    0.06 |  843.7500 | 578.1250 | 234.3750 | 4614.13 KB |        0.67 |

</details>

<details>
<summary>Benchmark results - Creation</summary>

| Method    | Job      | Runtime  | NumberOfFiles | LatencyMS | Mean             | Error         | StdDev        | Median           | Ratio | RatioSD | Gen0    | Gen1   | Allocated | Alloc Ratio |
|---------- |--------- |--------- |-------------- |---------- |-----------------:|--------------:|--------------:|-----------------:|------:|--------:|--------:|-------:|----------:|------------:|
| **Benchmark** | **Baseline** | **.NET 8.0** | **2**             | **-1**        |         **17.17 μs** |      **0.289 μs** |      **0.256 μs** |         **17.09 μs** |  **1.00** |    **0.00** |  **0.5798** |      **-** |   **2.44 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 2             | -1        |         14.20 μs |      0.214 μs |      0.179 μs |         14.26 μs |  0.83 |    0.01 |  0.5188 |      - |   2.15 KB |        0.88 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **2**             | **0**         |         **48.28 μs** |      **0.883 μs** |      **2.115 μs** |         **47.69 μs** |  **1.00** |    **0.00** |  **0.5493** |      **-** |   **2.48 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 2             | 0         |         20.52 μs |      0.390 μs |      0.848 μs |         20.22 μs |  0.42 |    0.03 |  0.5188 |      - |   2.19 KB |        0.88 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **2**             | **1**         |  **1,153,370.77 μs** |  **3,033.808 μs** |  **2,837.826 μs** |  **1,153,076.60 μs** |  **1.00** |    **0.00** |       **-** |      **-** |   **3.24 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 2             | 1         |    202,163.82 μs |    810.531 μs |    758.172 μs |    202,433.70 μs |  0.18 |    0.00 |       - |      - |   2.43 KB |        0.75 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **2**             | **5**         |  **1,154,271.62 μs** |  **2,629.954 μs** |  **2,460.061 μs** |  **1,153,689.70 μs** |  **1.00** |    **0.00** |       **-** |      **-** |   **3.24 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 2             | 5         |    202,071.53 μs |    987.541 μs |    923.747 μs |    201,793.43 μs |  0.18 |    0.00 |       - |      - |   2.43 KB |        0.75 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **10**            | **-1**        |         **59.75 μs** |      **0.906 μs** |      **0.803 μs** |         **59.52 μs** |  **1.00** |    **0.00** |  **3.2959** |      **-** |   **13.5 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 10            | -1        |         59.42 μs |      0.269 μs |      0.238 μs |         59.40 μs |  0.99 |    0.01 |  2.9297 |      - |  12.21 KB |        0.90 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **10**            | **0**         |        **226.82 μs** |      **5.396 μs** |     **15.395 μs** |        **227.63 μs** |  **1.00** |    **0.00** |  **3.1738** |      **-** |  **13.54 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 10            | 0         |         98.33 μs |      1.223 μs |      1.084 μs |         98.45 μs |  0.48 |    0.03 |  2.9297 |      - |  12.25 KB |        0.90 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **10**            | **1**         |  **5,274,439.79 μs** |  **6,901.351 μs** |  **6,455.528 μs** |  **5,273,943.90 μs** |  **1.00** |    **0.00** |       **-** |      **-** |  **14.26 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 10            | 1         |    950,228.86 μs |  2,438.419 μs |  2,161.595 μs |    950,750.30 μs |  0.18 |    0.00 |       - |      - |  12.97 KB |        0.91 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **10**            | **5**         |  **5,276,480.03 μs** |  **5,844.084 μs** |  **5,466.560 μs** |  **5,279,086.70 μs** |  **1.00** |    **0.00** |       **-** |      **-** |  **14.26 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 10            | 5         |    949,421.14 μs |  2,346.035 μs |  2,194.483 μs |    948,990.50 μs |  0.18 |    0.00 |       - |      - |  12.97 KB |        0.91 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **25**            | **-1**        |        **148.85 μs** |      **2.834 μs** |      **3.264 μs** |        **149.14 μs** |  **1.00** |    **0.00** |  **7.5684** |      **-** |  **31.16 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 25            | -1        |        147.24 μs |      1.313 μs |      1.229 μs |        146.78 μs |  0.99 |    0.03 |  6.8359 |      - |     28 KB |        0.90 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **25**            | **0**         |        **490.65 μs** |      **5.097 μs** |      **4.257 μs** |        **491.34 μs** |  **1.00** |    **0.00** |  **6.8359** |      **-** |   **31.2 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 25            | 0         |        214.95 μs |      1.275 μs |      1.065 μs |        214.92 μs |  0.44 |    0.00 |  6.8359 |      - |  28.04 KB |        0.90 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **25**            | **1**         | **13,010,754.08 μs** |  **7,587.337 μs** |  **7,097.199 μs** | **13,008,804.00 μs** |  **1.00** |    **0.00** |       **-** |      **-** |  **31.92 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 25            | 1         |  2,353,101.94 μs |  4,536.628 μs |  3,788.290 μs |  2,352,594.90 μs |  0.18 |    0.00 |       - |      - |  28.76 KB |        0.90 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **25**            | **5**         | **13,008,020.51 μs** |  **8,158.693 μs** |  **7,631.646 μs** | **13,008,868.30 μs** |  **1.00** |    **0.00** |       **-** |      **-** |  **31.92 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 25            | 5         |  2,356,024.47 μs |  4,256.167 μs |  3,981.221 μs |  2,356,131.80 μs |  0.18 |    0.00 |       - |      - |  28.76 KB |        0.90 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **50**            | **-1**        |        **294.44 μs** |      **1.924 μs** |      **1.800 μs** |        **294.02 μs** |  **1.00** |    **0.00** | **15.1367** |      **-** |   **62.7 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 50            | -1        |        297.19 μs |      4.039 μs |      3.580 μs |        297.07 μs |  1.01 |    0.01 | 13.6719 | 0.9766 |  56.41 KB |        0.90 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **50**            | **0**         |        **981.38 μs** |      **4.847 μs** |      **4.534 μs** |        **980.08 μs** |  **1.00** |    **0.00** | **13.6719** |      **-** |  **62.74 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 50            | 0         |        428.33 μs |      3.014 μs |      2.672 μs |        428.36 μs |  0.44 |    0.00 | 13.6719 | 0.4883 |  56.45 KB |        0.90 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **50**            | **1**         | **25,900,673.22 μs** | **15,093.982 μs** | **14,118.920 μs** | **25,902,252.70 μs** |  **1.00** |    **0.00** |       **-** |      **-** |  **63.46 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 50            | 1         |  4,699,400.51 μs |  5,271.756 μs |  4,931.204 μs |  4,697,546.80 μs |  0.18 |    0.00 |       - |      - |  57.17 KB |        0.90 |
|           |          |          |               |           |                  |               |               |                  |       |         |         |        |           |             |
| **Benchmark** | **Baseline** | **.NET 8.0** | **50**            | **5**         | **25,891,853.07 μs** | **12,417.690 μs** | **11,615.515 μs** | **25,891,875.90 μs** |  **1.00** |    **0.00** |       **-** |      **-** |  **63.46 KB** |        **1.00** |
| Benchmark | CoreRun  | .NET 9.0 | 50            | 5         |  4,698,669.15 μs |  7,373.175 μs |  6,896.872 μs |  4,699,045.70 μs |  0.18 |    0.00 |       - |      - |  57.17 KB |        0.90 |

</details>

<details>
<summary>Benchmark header</summary>

```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3593/23H2/2023Update/SunValley3)
Intel Core i7-8565U CPU 1.80GHz (Whiskey Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.200
  [Host]   : .NET 8.0.6 (8.0.624.26715), X64 RyuJIT AVX2
  Baseline : .NET 8.0.6 (8.0.624.26715), X64 RyuJIT AVX2
  CoreRun  : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
```

</details>

NB: because this changes the number of reads/writes to a stream, it'll have an impact on the tests for #102704. I'll change these depending on the order the PRs are merged in.

edwardneal committed:
Removed BinaryReader, BinaryWriter from ZipArchive and ZipBlocks

We now read the data into a (sometimes stack-allocated) byte array and slice it up with BinaryPrimitives.
This reduces the number of reads and writes required to read and write a ZipArchive. It also makes future work to enable async APIs easier, since BinaryReader and BinaryWriter lack this support.
Also changed approach to reading central file directory headers. Rather than performing X reads per header, we read 4KB of data at a time and look for all applicable headers in that data. This should improve performance when dealing with many small files.

edwardneal committed:
Removed BinaryReader from ZipArchiveEntry

This allowed the removal of the ArchiveReader property from ZipArchive.

edwardneal committed:
Removed BinaryWriter from ZipArchiveEntry

edwardneal committed:
Re-adding missing assertion

edwardneal committed:
Reduced memory usage with array pooling

Now pooling the file IO buffers and the temporary buffers for extra fields of the CD file header (which would otherwise be allocated and deallocated in a loop.)

bartonjs commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -524,28 +523,78 @@ internal void WriteCentralDirectoryFileHeader()
+            // Version made by Specification (version)  (1 byte)
+            cdStaticHeader[cdPosition++] = (byte)_versionMadeBySpecification;
+
+            // Version made by Compatibility (type)     (1 byte)
+            cdStaticHeader[cdPosition++] = (byte)CurrentZipPlatform;                        // Version made by Compatibility (type)     (1 byte)
```
The comment doesn't need to be there twice, at minimum.

```suggestion
            // Version made by Compatibility (type)     (1 byte)
            cdStaticHeader[cdPosition++] = (byte)CurrentZipPlatform;
```

edwardneal replied:
Thanks - I missed that comment for some reason, but have checked the rest; it's the only one.

edwardneal committed:
Corrected duplicate comment

edwardneal committed:
Added test for changed central directory read method

This handles 64x entries with 19-character filenames (and thus, 65-byte file headers.) As a result, it straddles two 4KB read buffers.

Also corrected the seek logic while reading the central directory header

carlossanlop commented:
@edwardneal the `main` branch is currently only taking changes for RC1 that will go into .NET 9. This change currently does not meet the bar for merging it now, but we would love to consider taking it for .NET 10. The RC1 branch will get snapped from `main` on August 14th, and we can consider merging this PR after that date, when `main` will start pointing to .NET 10.

carlossanlop commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs:
```diff
@@ -273,6 +273,37 @@ public static async Task EnsureDisposeIsCalledAsExpectedOnTheUnderlyingStream(bo
             Assert.Equal(expectedDisposeCalls, disposeCallCountingStream.NumberOfDisposeCalls);
         }
 
+        [Theory]
+        // A 19-character filename will result in a 65-byte central directory header. 64 of these will make the central directory
+        // read process stretch into two 4KB buffers.
+        [InlineData(64, "example/file-{0:00}.dat")]
+        public static void CanReadLargeCentralDirectoryHeader(int count, string entryNameFormat)
+        {
+            using (MemoryStream archiveStream = new MemoryStream())
```
If there's only one `InlineData`, we can convert this to a `Fact`.

```suggestion
        // A 19-character filename will result in a 65-byte central directory header. 64 of these will make the central directory
        // read process stretch into two 4KB buffers.
        [Fact]
        public static void CanReadLargeCentralDirectoryHeader()
        {
            int count = 64;
            string entryNameFormat = "example/file-{0:00}.dat";

            using (MemoryStream archiveStream = new MemoryStream())
```

carlossanlop reviewed (commented):
A quick partial review. Also needs resolving the merge conflict so we get proper CI results.

edwardneal committed:
Merge remote-tracking branch 'upstream/main' into ziparchive-stream-read-write

edwardneal committed:
Test changes: formatting, converted Theory to Fact

edwardneal commented:
Thanks @carlossanlop - I've addressed the merge conflict and made those test changes.

I've got a number of post-merge test failures to deal with - I'll look at those shortly, ready for review post-14th.

edwardneal committed:
Resolving test failures

The buffer returned from the ArrayPool contained older data (including a ZIP header.) When reading the last chunk of the file (i.e a chunk which was less than BackwardsSeekingBufferSize) the buffer's Span wasn't resized to account for this.
SeekBackwardsToSignature would thus find the older data, try to seek beyond the end of the stream and fail to read the file.

carlossanlop committed:
Merge branch 'main' into ziparchive-stream-read-write

carlossanlop commented:
I fetched your branch and all the relevant tests passed locally: System.IO.Compression, System.IO.Compression.ZipFile and System.IO.Packaging.

I merged the latest bits in main into this branch, as the last CI results were already stale and the build info was gone. Let's see what comes out of this.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -483,32 +477,90 @@ private void EnsureCentralDirectoryRead()
             try
             {
-                // assume ReadEndOfCentralDirectory has been called and has populated _centralDirectoryStart
+                // assume ReadEndOfCentralDirectory has been called and has populated _centralDirectoryStart and _expectedCentralDirectorySize
```
I don't see `_expectedCentralDirectorySize` being used anywhere. It only gets values assigned to it.

carlossanlop replied:
What were you intending to do with it? Does it have the purpose of just verifying that an initial valid value was already read for the central directory?

If that's the case, then I was thinking that since `_expectedCentralDirectorySize` only acquires a valid value _after_ `ReadEndOfCentralDirectory` is called, and `ReadEndOfCentralDirectory` is called in the `ZipArchive` constructor, then at the very top of the `ZipArchive` constructor, after the `ThrowIfNull` checks, you could set the initial value of `_expectedCentralDirectorySize` to `-1`:

https://github.com/dotnet/runtime/blob/e1e3cdccb872327488f8d38332f8caa540ce4d03/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs#L121-L123

and then here, right under this new comment, you could add a debug assert check to confirm it already acquired a valid value:

```suggestion
                // assume ReadEndOfCentralDirectory has been called and has populated _centralDirectoryStart and _expectedCentralDirectorySize
                Debug.Assert(_expectedCentralDirectorySize != -1);
```

But if that was not your intention with this field, then consider removing it.

edwardneal replied:
I should have removed `_expectedCentralDirectorySize` before submitting the PR, apologies. The original intent was twofold:
1. Use the expected size to validate the central directory as we read it. We currently validate that the file has the expected number of records in the central directory, but we don't validate the total size of the central directory.
2. Rather than performing X reads from the source stream per CD entry, consider reading `_expectedCentralDirectorySize` bytes from the stream and parsing the list of entries in memory.

I decided against doing so:
1. I wanted to avoid the breaking change for point 1.
2. Point 2 had the potential to balloon memory usage for a maliciously-constructed ZIP file. I could have overcome that by reading in 4KB chunks, but by the time I was thinking about handling central directory entries which straddled chunks, I felt that was pushing the scope of the PR.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -526,23 +578,23 @@ private void ReadEndOfCentralDirectory()
                 // But if the EOCD has max possible size, the signature should be found somewhere in the previous 64K + 4 bytes
                 if (!ZipHelper.SeekBackwardsToSignature(_archiveStream,
-                        ZipEndOfCentralDirectoryBlock.SignatureConstant,
-                        ZipEndOfCentralDirectoryBlock.ZipFileCommentMaxLength + ZipEndOfCentralDirectoryBlock.SignatureSize))
```
Seems that after your changes, `ZipEndOfCentralDirectoryBlock.SignatureSize` is no longer used anywhere. Can you please remove it?

edwardneal replied:
It's gone, thanks!

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -536,28 +535,78 @@ internal void WriteCentralDirectoryFileHeader()
+            BinaryPrimitives.WriteUInt32LittleEndian(cdStaticHeader.Slice(cdPosition), compressedSizeTruncated);
+            cdPosition += sizeof(uint);
+
+            // Uncompressed Size                        (4 bytes)
+            BinaryPrimitives.WriteUInt32LittleEndian(cdStaticHeader.Slice(cdPosition), uncompressedSizeTruncated);
+            cdPosition += sizeof(uint);
+
+            // File Name Length                         (2 bytes)
+            BinaryPrimitives.WriteUInt16LittleEndian(cdStaticHeader.Slice(cdPosition), (ushort)_storedEntryNameBytes.Length);
+            cdPosition += sizeof(ushort);
+
+            // Extra Field Length                       (2 bytes)
+            BinaryPrimitives.WriteUInt16LittleEndian(cdStaticHeader.Slice(cdPosition), extraFieldLength);
+            cdPosition += sizeof(ushort);
+
+            // File Comment Length                      (2 bytes)
             Debug.Assert(_fileComment.Length <= ushort.MaxValue);
+            BinaryPrimitives.WriteUInt16LittleEndian(cdStaticHeader.Slice(cdPosition), (ushort)_fileComment.Length);
+            cdPosition += sizeof(ushort);
+
+            // Start Disk Number                        (2 bytes)
+            BinaryPrimitives.WriteUInt16LittleEndian(cdStaticHeader.Slice(cdPosition), 0);
+            cdPosition += sizeof(ushort);
+
+            // Internal File Attributes                 (2 bytes)
+            BinaryPrimitives.WriteUInt16LittleEndian(cdStaticHeader.Slice(cdPosition), 0);
+            cdPosition += sizeof(ushort);
+
+            // External File Attributes                 (4 bytes)
+            BinaryPrimitives.WriteUInt32LittleEndian(cdStaticHeader.Slice(cdPosition), _externalFileAttr);
+            cdPosition += sizeof(uint);
 
-            writer.Write((ushort)_fileComment.Length);
-            writer.Write((ushort)0); // disk number start
-            writer.Write((ushort)0); // internal file attributes
-            writer.Write(_externalFileAttr); // external file attributes
-            writer.Write(offsetOfLocalHeaderTruncated); // offset of local header
+            // Offset Of Local Header                   (4 bytes)
+            BinaryPrimitives.WriteUInt32LittleEndian(cdStaticHeader.Slice(cdPosition), offsetOfLocalHeaderTruncated);
+            cdPosition += sizeof(uint);
```
This is a great contribution. I love this.

I did something similar for System.Formats.Tar, but instead of making addition operations for the `cdPosition` every time the next field had to be written, I opted for storing all the expected field sizes and field positions, and use those instead to avoid making operations every time:

[System.Formats.Tar/src/System/Formats/Tar/FieldLengths.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Formats.Tar/src/System/Formats/Tar/FieldLengths.cs)
[System.Formats.Tar/src/System/Formats/Tar/FieldLocations.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Formats.Tar/src/System/Formats/Tar/FieldLocations.cs)

Then each time we have to read or write data, those constant fields are used alongside span slice to find the right place to read or write the data.

Can you please consider doing the same? I think having fields locations and lengths make the code clearer to understand and maintain, while live operations can lead to mistakes that are difficult to investigate or track if we ever need to modify this code.

Reading:
https://github.com/dotnet/runtime/blob/7ddd2f6dea8a1aa2eb3e97750831cbedb3e25e6a/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs#L386-L397

Writing:
https://github.com/dotnet/runtime/blob/7ddd2f6dea8a1aa2eb3e97750831cbedb3e25e6a/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs#L610-L662

edwardneal replied:
I've swapped these over, and I see what you mean thanks. These are quite a bit more readable.

I've moved the field lengths and positions to two dedicated ZipBlocks.*.cs files. Many of the members here are declared as `static readonly` variables, because they indirectly reference the length of an RVA span (`SignatureConstantBytes.Length`). This [JITs](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUEZrgJwAUAKgKYoIB0t9ASgGoCCDAMjTAOYIAFgEoA3ERIBmAAQ4ZzBEQDeRGermzSANhksaEACYB5GABsAngGUADhBgAeAEIWENAHwK6CdhxkBeTwAiAAkaMzMAezgZAHdIgCczQwBCIIBXAA4JQgBfIA==) to a constant value, but it isn't ideal at compile-time.

The current approach is safest - the various `FieldLengths.Signature` members directly reference `SignatureConstantBytes.Length`, so there's no way for the signature values and their lengths to fall out of sync. These values should always match the constant values in the spec. though, so if you'd prefer for their lengths to be compile-time constants then I'm happy to change them.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -938,19 +988,40 @@ private bool WriteLocalFileHeader(bool isEmptyFile)
+            currOffset += sizeof(ushort);
+
+            BinaryPrimitives.WriteUInt16LittleEndian(lfStaticHeader.Slice(currOffset), extraFieldLength);
+            currOffset += sizeof(ushort);
```
Here and elsewhere where a read or a write is done, please consider my suggestion of using constant field lengths and positions where they apply.

Obviously there are cases where the data length and position is variable, so that's where it's ok to do an operation.

carlossanlop reviewed (commented):
Left more feedback for you to consider. Thank you so much for this change.

edwardneal committed:
Responded to code review

Reads and writes are now performed using a new set of field lengths and locations.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs:
```diff
@@ -103,91 +103,85 @@ internal static uint DateTimeToDosTime(DateTime dateTime)
-                while (bufferPointer >= 0 && !signatureFound)
+                while (!signatureFound && !outOfBytes && totalBytesRead <= maxBytesToRead)
                 {
-                    currentSignature = (currentSignature << 8) | ((uint)buffer[bufferPointer]);
```
Glad we're getting rid of this line. Not a fan of bit shifting operations when there are easier-to-read alternatives.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs:
```diff
@@ -103,91 +103,85 @@ internal static uint DateTimeToDosTime(DateTime dateTime)
 
-                Debug.Assert(bufferPointer < buffer.Length);
+                int totalBytesRead = 0;
+                int duplicateBytesRead = 0;
```
Could you share an explanation of why this variable is named `duplicateBytesRead`? I wonder if we could use a different name. I'm not sure I understand what duplication this variable is referring to. Maybe with the explanation I could understand it. And maybe a comment above it would suffice.

edwardneal replied:
`SeekBackwardsToSignature` is a little unusual: it reads blocks of 4096 bytes at a time into `bufferSpan`, but then seeks backwards by 4092 bytes at a time. This is where the idea of an overlap in the `SeekBackwardsAndRead` comes in: it's the number of bytes which will appear in bufferSpan in two successive reads. These are the bytes which I describe as "duplicate."

I introduced this difference between the two sizes because it guarantees that if the 4-byte signature exists in the file, it'll never "straddle" two successive reads (where the first read begins with bytes 3 & 4 and the second read ends with bytes 1 & 2.) As a result, the bitshifting can be replaced completely with a simpler, vectorised, LastIndexOf.

The tradeoff is that we read four bytes more than we strictly require per iteration... but this is a 0.1% inefficiency, and it still shakes out as faster than the current implementation.

carlossanlop replied:
Thanks for the explanation.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -526,16 +577,15 @@ private void ReadEndOfCentralDirectory()
-                        ZipEndOfCentralDirectoryBlock.SignatureConstant,
-                        ZipEndOfCentralDirectoryBlock.ZipFileCommentMaxLength + ZipEndOfCentralDirectoryBlock.SignatureSize))
+                        ZipEndOfCentralDirectoryBlock.SignatureConstantBytes,
+                        ZipEndOfCentralDirectoryBlock.ZipFileCommentMaxLength + ZipEndOfCentralDirectoryBlock.SignatureConstantBytes.Length))
```
Here and elsewhere: I believe you should now use `*.FieldLengths.Signature` instead of `*.SignatureConstantBytes.Length`.

edwardneal replied:
Yes, sorry. I didn't see these review comments - I'll roll them into the next commit with the remaining review!

carlossanlop replied:
No worries. Ping me when you're done with all your updates and I'll run the mobile platforms leg.

edwardneal replied:
Thanks @carlossanlop. The main CI leg is running, and I'm ready for the mobile platforms leg.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -587,14 +637,12 @@ private void TryReadZip64EndOfCentralDirectory(ZipEndOfCentralDirectoryBlock eoc
-                        Zip64EndOfCentralDirectoryLocator.SignatureConstant,
-                        Zip64EndOfCentralDirectoryLocator.SignatureSize))
+                        Zip64EndOfCentralDirectoryLocator.SignatureConstantBytes,
+                        Zip64EndOfCentralDirectoryLocator.SignatureConstantBytes.Length))
```
`Zip64EndOfCentralDirectoryLocator.FieldLengths.Signature`

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:
```diff
@@ -358,107 +439,160 @@ internal struct Zip64EndOfCentralDirectoryRecord
+            int bytesRead = stream.Read(blockBytes);
 
-            if (reader.ReadUInt32() != SignatureConstant)
+            if (bytesRead != SignatureConstantBytes.Length || !blockBytes.SequenceEqual(SignatureConstantBytes))
```
`*.FieldLengths.Signature`, but which one?

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -536,28 +538,46 @@ internal void WriteCentralDirectoryFileHeader()
+            BinaryPrimitives.WriteUInt16LittleEndian(cdStaticHeader[ZipCentralDirectoryFileHeader.FieldLocations.InternalFileAttributes..], 0);
+            BinaryPrimitives.WriteUInt32LittleEndian(cdStaticHeader[ZipCentralDirectoryFileHeader.FieldLocations.ExternalFileAttributes..], _externalFileAttr);
+            BinaryPrimitives.WriteUInt32LittleEndian(cdStaticHeader[ZipCentralDirectoryFileHeader.FieldLocations.RelativeOffsetOfLocalHeader..], offsetOfLocalHeaderTruncated);
+
```
Nice!

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -1037,31 +1066,39 @@ private void WriteCrcAndSizesInLocalHeader(bool zip64HeaderUsed)
-                writer.Write(_crc32);
-                writer.Write(compressedSizeTruncated);
-                writer.Write(uncompressedSizeTruncated);
+                int relativeCrc32Location = ZipLocalFileHeader.FieldLocations.Crc32 - ZipLocalFileHeader.FieldLocations.Crc32;
+                int relativeCompressedSizeLocation = ZipLocalFileHeader.FieldLocations.CompressedSize - ZipLocalFileHeader.FieldLocations.Crc32;
+                int relativeUncompressedSizeLocation = ZipLocalFileHeader.FieldLocations.UncompressedSize - ZipLocalFileHeader.FieldLocations.Crc32;
```
I think these could be `const`, like the variables declared at the top of the method.

edwardneal replied:
They can't at the moment - many of the FieldLocations members are `readonly static` variables. I can't make those `const` because they all ultimately reference to FieldLengths.Signature. FieldLengths.Signature, in turn, references the length of SignatureConstantBytes - and that's not a compile-time constant.

It'd be good to make this more consistent, but I'd need to change `FieldLengths.Signature` to be a magic number of 4 and we'd lose a bit of clarity because we're no longer directly referencing `SignatureConstantBytes.Length`. I could do that and add a comment though?

From running this through sharplab, they both shake out as JIT-time constants - it's just a style choice between clarity and consistency...

carlossanlop replied:
Ah got it. Nevermind then.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -1037,31 +1066,39 @@ private void WriteCrcAndSizesInLocalHeader(bool zip64HeaderUsed)
             // and setting the version to Zip64 to indicate that descriptor contains 64-bit values
             if (pretendStreaming)
             {
+                int relativeVersionToExtractLocation = ZipLocalFileHeader.FieldLocations.VersionNeededToExtract - ZipLocalFileHeader.FieldLocations.VersionNeededToExtract;
+                int relativeGeneralPurposeBitFlagsLocation = ZipLocalFileHeader.FieldLocations.GeneralPurposeBitFlags - ZipLocalFileHeader.FieldLocations.VersionNeededToExtract;
```
I think these could be `const`, like the variables declared at the top of the method.

carlossanlop replied:
Nevermind. If not possible, it's ok te leave them as is.

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:
```diff
@@ -180,82 +204,99 @@ private static bool TryGetZip64BlockFromGenericExtraField(ZipGenericExtraField e
+            }
+
+            return true;
+
```
```suggestion
```

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs:
```diff
@@ -299,53 +340,93 @@ public static void RemoveZip64Blocks(List<ZipGenericExtraField> extraFields)
     {
-        public const uint SignatureConstant = 0x07064B50;
-        public const int SignatureSize = sizeof(uint);
+        public static ReadOnlySpan<byte> SignatureConstantBytes => [0x50, 0x4B, 0x06, 0x07];
```
I liked your comment above the other signature. Can you add it here too, please?:
```suggestion
        // The Zip File Format Specification references 0x07064B50, this is a big endian representation.
        // ZIP files store values in little endian, so this is reversed.
        public static ReadOnlySpan<byte> SignatureConstantBytes => [0x50, 0x4B, 0x06, 0x07];
```

carlossanlop reviewed (commented):
@edwardneal thanks for applying the suggestions, I'm happy with how this is looking! I left a (hopefully) final batch of comments for you to consider.

After addressing those comments, I'd like to run a CI run that doesn't get executed by default, which will test your code in mobile platforms. If we don't see anything concerning there, I think we can merge it.

If it's in your possibilities, we could try to get this merged before the Code Complete day for .NET 10 Preview 1, which is Monday January 27th. Let me know if this works for you, otherwise I can take over for the above two final steps.

edwardneal committed:
Code review response

Formatting change; added one comment to Zip64EndOfCentralDirectoryLocator.SignatureConstantBytes; clarified comment on ZipHelper.SeekBackwardsAndRead

edwardneal commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs:
```diff
@@ -103,91 +103,85 @@ internal static uint DateTimeToDosTime(DateTime dateTime)
-        internal static void AdvanceToPosition(this Stream stream, long position)
+        // Returns the number of bytes actually read.
+        // Allows successive buffers to overlap by a number of bytes (to handle cases where the
+        // value being searched for straddles buffers.)
```
It doesn't read very well. I think I've made it clearer now. The comment's trying to say that we let successive buffers overlap by X bytes to handle situations where the first buffer ends with the first few bytes of the ZIP signature, and the second buffer starts with the remaining bytes of that signature.

carlossanlop replied:
Oooh I got it now. My English wasn't very good yesterday (needed coffee).

edwardneal commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs:
```diff
@@ -103,91 +103,85 @@ internal static uint DateTimeToDosTime(DateTime dateTime)
-        {
             if (stream.Position >= buffer.Length)
             {
+                stream.Seek(-(buffer.Length - overlap), SeekOrigin.Current);
```
`overlap` is the length of the ZIP signature (which is constant) and `buffer` is always a constant number of bytes long, so I don't think the condition you're describing can ever happen. We should make that explicit though, so I agree - I've added the assert.

edwardneal commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs:
```diff
@@ -103,91 +103,85 @@ internal static uint DateTimeToDosTime(DateTime dateTime)
             else
             {
                 int bytesToRead = (int)stream.Position;
                 stream.Seek(0, SeekOrigin.Begin);
```
It's essentially a `Math.Min`. To reach this point, the stream's position must be smaller than the buffer size (4KB, in this case.) In that situation, we can't seek backwards by the buffer size because it'll take us to a position before the start of the stream. Instead, we just go to the start of the stream and read `stream.Position` bytes.

The second Seek is there to duplicate the behaviour of the other code path: at the end of SeekBackwardsAndRead, the stream's position is always at the start of the Stream contents read into `buffer`.

carlossanlop replied:
Okay, I think I got it.

edwardneal committed:
Additional code review comments

Added comment to SeekBackwardsToStream.
Lingering references to SignatureConstantBytes.Length.
Added two asserts to CanReadLargeCentralDirectoryHeader test, verifying that the archive entry metadata is in a sensible state.

edwardneal commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs:
```diff
@@ -273,6 +273,38 @@ public static async Task EnsureDisposeIsCalledAsExpectedOnTheUnderlyingStream(bo
+
+                    for (int i = 0; i < count; i++)
+                    {
+                        Assert.Equal(string.Format(entryNameFormat, i), readArchive.Entries[i].FullName);
```
I've added two asserts on CompressedSize and UncompressedSize - the main point of the test is to prove that the new central directory read logic can find every entry, but it doesn't hurt to make sure other fields are as expected.

carlossanlop reviewed (approved):
LGTM! Thank you @edwardneal for this PR.

The CI failures in the default legs are all marked as expected. The runtime-extra-platforms legs unfortunately are plagued with `exit 137` for all assemblies, so we'll have to verify the results there at a later time after that issue gets resolved.

edwardneal commented:
@carlossanlop I've seen several outerloop tests failing on `main` when dealing with Zip64 structures. I'm pretty sure they came from this PR and I'm going to trace them down this evening - sorry for missing these.

carlossanlop commented:
Okay, thank you. We might have to revert both to avoid risking the preview.
